### PR TITLE
3947 support http delete for operation $export poll status

### PR DESCRIPTION
**Pull request by StevenXLi**
- Added a new parameter to operations to support the DELETE request type
- Added support for the new parameter in OperationMethodBinding to bind to poll status operation
- Added new method and api to Batch2JobRunner, to expose cancel function from batch2
  - Also added new Object `Batch2JobOperationResult` for the return type of this
- Modified `exportPollStatus` operation implementation in BulkDataExportProvider to support the DELETE request type

### CODE REVIEW

**Review by jmarchionatto — APPROVED** [review]
Neat work!

File: hapi-fhir-storage/src/main/java/ca/uhn/fhir/jpa/bulk/export/provider/BulkDataExportProvider.java [hunk 1]
```diff
@@ -299,17 +303,35 @@ public void exportPollStatus(
...
 				break;
 		}
 	}
 
+	private void handleDeleteRequest(IPrimitiveType<String> theJobId, HttpServletResponse response, BulkExportJobStatusEnum theStatusEnum) throws IOException {
+		IBaseOperationOutcome outcome = OperationOutcomeUtil.newInstance(myFhirContext);
+		Batch2JobOperationResult resultMessage = myJobRunner.cancelInstance(theJobId.getValueAsString());
+		if (theStatusEnum.equals(BulkExportJobStatusEnum.COMPLETE)) {
+			response.setStatus(Constants.STATUS_HTTP_404_NOT_FOUND);
+			OperationOutcomeUtil.addIssue(myFhirContext, outcome, "error", "Job instance <" + theJobId.getValueAsString() + "> was already cancelled.  Nothing to do.", null, null);
```
[the remark this answers is not in the document]
Comment (StevenXLi) [reply hapi-fhir-storage/src/main/java/ca/uhn/fhir/jpa/bulk/export/provider/BulkDataExportProvider.java:326]: So `BulkExportJobStatusEnum` doesn't have a cancelled status, the comments and code suggests that it's translated to ERROR, but from testing, after cancelling a bulk job, the status is set to COMPLETE. Verified with the binaries not even generated.

[the remark this answers is not in the document]
Comment (jmarchionatto) [reply hapi-fhir-storage/src/main/java/ca/uhn/fhir/jpa/bulk/export/provider/BulkDataExportProvider.java:326]: What I meant is that the status COMPLETE doesn't guarantee you the job was cancelled, so the message could be misleading if the job was completed normally.

[the remark this answers is not in the document]
Comment (StevenXLi) [reply hapi-fhir-storage/src/main/java/ca/uhn/fhir/jpa/bulk/export/provider/BulkDataExportProvider.java:326]: Would changing it to `Job instance ... was already cancelled or has completed.  Nothing to do.` be better?
It has to be a error however, as part of the requirements.

[the remark this answers is not in the document]
Comment (jmarchionatto) [reply hapi-fhir-storage/src/main/java/ca/uhn/fhir/jpa/bulk/export/provider/BulkDataExportProvider.java:326]: Yes. I find that more clear

File: hapi-fhir-jpaserver-test-utilities/src/test/java/ca/uhn/fhir/jpa/bulk/BulkDataExportProviderTest.java [hunk 2]
```diff
@@ -676,6 +680,74 @@ public void testProviderReturnsSameIdForSameJob() throws IOException {
...
+
+			assertEquals(202, response.getStatusLine().getStatusCode());
+			assertEquals("Accepted", response.getStatusLine().getReasonPhrase());
+
+			verify(myJobRunner, times(1)).cancelInstance(A_JOB_ID);
+			String responseContent = IOUtils.toString(response.getEntity().getContent(), Charsets.UTF_8);
+			ourLog.info("Response content: {}", responseContent);
+			if (!Objects.equals(responseContent, "")) {
+				assertThat(responseContent, containsString("successfully cancelled."));
+			}
```
[the remark this answers is not in the document]
Comment (StevenXLi) [reply hapi-fhir-jpaserver-test-utilities/src/test/java/ca/uhn/fhir/jpa/bulk/BulkDataExportProviderTest.java]: removed

File: hapi-fhir-jpaserver-test-utilities/src/test/java/ca/uhn/fhir/jpa/bulk/BulkDataExportProviderTest.java [hunk 3]
```diff
@@ -676,6 +680,74 @@ public void testProviderReturnsSameIdForSameJob() throws IOException {
...
+
+			assertEquals(404, response.getStatusLine().getStatusCode());
+			assertEquals("Not Found", response.getStatusLine().getReasonPhrase());
+
+			verify(myJobRunner, times(1)).cancelInstance(A_JOB_ID);
+			String responseContent = IOUtils.toString(response.getEntity().getContent(), Charsets.UTF_8);
+			ourLog.info("Response content: {}", responseContent);
+			if (!Objects.equals(responseContent, "")) {
+				assertThat(responseContent, containsString("was already cancelled."));
+			}
```
[the remark this answers is not in the document]
Comment (StevenXLi) [reply hapi-fhir-jpaserver-test-utilities/src/test/java/ca/uhn/fhir/jpa/bulk/BulkDataExportProviderTest.java]: also removed

File: hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/method/OperationMethodBinding.java [hunk 4]
```diff
@@ -312,20 +315,29 @@ public Object invokeServer(IRestfulServer<?> theServer, RequestDetails theReques
...
+			allowedRequestTypes.add(RequestTypeEnum.DELETE);
+		}
+		String messageParameter = allowedRequestTypes.stream().map(RequestTypeEnum::name).collect(Collectors.joining(", "));
 		if (theRequest.getRequestType() == RequestTypeEnum.POST) {
 			// all good
 		} else if (theRequest.getRequestType() == RequestTypeEnum.GET) {
 			if (!myIdempotent) {
-				String message = getContext().getLocalizer().getMessage(OperationMethodBinding.class, "methodNotSupported", theRequest.getRequestType(), RequestTypeEnum.POST.name());
-				throw new MethodNotAllowedException(Msg.code(426) + message, RequestTypeEnum.POST);
+				String message = getContext().getLocalizer().getMessage(OperationMethodBinding.class, "methodNotSupported", theRequest.getRequestType(), messageParameter);
```
[the remark this answers is not in the document]
Comment (StevenXLi) [reply hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/method/OperationMethodBinding.java]: factored it out, thanks for catching this

File: hapi-fhir-storage/src/main/java/ca/uhn/fhir/jpa/bulk/export/provider/BulkDataExportProvider.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (StevenXLi) [reply hapi-fhir-storage/src/main/java/ca/uhn/fhir/jpa/bulk/export/provider/BulkDataExportProvider.java]: changed

**Pull request merged**